src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/RedundantImportCheck.java

```diff
@@ -141,6 +149,31 @@ else if (pkgName != null && isFromPackage(imp.getText(), pkgName)) {
         }
     }
 
+    @Override
+    public void finishTree(DetailAST ast) {
+
+        // check for types already imported by star imports
+        // we should run this only if there is only 1 star import in file
+        // 0 star imports -- there will be no violations
+        // > 1 star imports -- false positives are possible
+        if (starImports.size() == 1) {
+            final FullIdent starImp = starImports.get(0);
+
+            imports.stream()
+                .filter(imp -> !isStarImport(imp))
+                .filter(imp -> isDuplicatedByStarImport(starImp, imp))
+                .forEach(imp -> {
+                    log(
+                        imp.getLineNo(),
+                        imp.getColumnNo() - imp.getText().length(),
```

Comment:
Violation should be the AST otherwise xpath will never be able to suppress this violation.
Is there some way we can give it the AST?

Reply:
@rnveach Is it ok if I define utility pair-like class inside RedundantImportCheck class and use it to store AST along with FullIdent?

Reply:
I should be fine with a utility class that is just a group of fields. Another way you could do this is with just a simple Map of FullIdent to AST since you only have 2 fields.
Whichever looks nicer and easier to understand is fine with me.